File: java/src/com/sortcompare/SelectionSort.java
package com.sortcompare;

import java.util.*;

//Performs the Selection Sort algorithm. Performance is O(n^2)

class SelectionSort {

    static ArrayList<Integer> sort(final ArrayList<Integer> unsortedList) {
        final ArrayList<Integer> result = new ArrayList<>();
        final int originalSize = unsortedList.size();
        for (int i = 0; i < originalSize; i++) {
            int smallestIndex = findSmallestIndex(unsortedList);
            result.add( unsortedList.remove(smallestIndex));
        }
        return result;

    }

    private static int findSmallestIndex(final ArrayList<Integer> unsortedList) {
        int smallest = unsortedList.get(0);
        int smallestIndex = 0;
        for (int i = 0; i < unsortedList.size(); i++) {
            if( unsortedList.get(i) < smallest) {
                smallest = unsortedList.get(i);
                smallestIndex = i;
            }
        }
        return smallestIndex;
    }
}

File: kotlin/src/sortcompare/SelectionSort.kt
package sortcompare

import java.util.*

class SelectionSort {

    companion object {
        fun sort(unsortedList: ArrayList<Int>): ArrayList<Int> {
            val sortedList: ArrayList<Int> = ArrayList()
            val originalSize = unsortedList.size
            for(i in 1..originalSize) {
                val smallestIndex = findSmallestIndex(unsortedList)
                val smallest = unsortedList.removeAt(smallestIndex)
                sortedList.add(smallest)
            }
            return sortedList
        }

        fun findSmallestIndex(unsortedList: ArrayList<Int>): Int {
            var smallest = unsortedList[0]
            var smallestIndex = 0
            unsortedList.forEachIndexed { index, i ->
                if (i < smallest) {
                    smallest = i
                    smallestIndex = index
                }
            }
            return smallestIndex
        }
    }
}